### DISCUSSION TITLE
rule for Sonoff TX Ultimate does not work

### DISCUSSION THREAD

**Discussion by darkdan2000** [arendst/Tasmota#22037] [Q&A]
Hi friends,

I have a problem with the rules for switching the relay on the Sonoff TX Ultimate (T5 4-Gang).

The rule ([found here](https://templates.blakadder.com/sonoff_T5-2C-86.html)) which is specified for the 2-Gang switch, for example, doesn't work for me.

`rule2 On TXultimate#Action$!Touch do break rule2 + On TXUltimate#Channel<5 do power1 toggle endon rule2 + On TXUltimate#Channel>5 do power2 toggle endon `

### Log output:
16:00:21.318 bytes('AA5501020200012874')
16:00:21.320 Touch event: 0 pos: 1
16:00:21.325 MQT: tele/Sonoff-Lichtschalter-Keller/RESULT = {"TXUltimate":{"Action":"Touch","Channel":1}}
16:00:21.518 bytes('AA5501020101BF04')
16:00:21.520 Short press zone: 1
16:00:21.525 MQT: tele/Sonoff-Lichtschalter-Keller/RESULT = {"TXUltimate":{"Action":"Short","Channel":1}}
**_16:00:21.533 RUL: TXULTIMATE#ACTION$!TOUCH performs ''_**

What works is:

`rule2 On TXultimate#Action=Touch do power1 toggle endon`

### Log Output:
15:43:25.151 bytes('AA5501020200012874')
15:43:25.153 Touch event: 0 pos: 1
15:43:25.159 MQT: tele/Sonoff-Lichtschalter-Keller/RESULT = {"TXUltimate":{"Action":"Touch","Channel":1}}
**_15:43:25.166 RUL: TXULTIMATE#ACTION=TOUCH performs 'power1 toggle'_**
15:43:25.171 MQT: stat/Sonoff-Lichtschalter-Keller/RESULT = {"POWER1":"ON"}
15:43:25.173 MQT: stat/Sonoff-Lichtschalter-Keller/POWER1 = ON
15:43:25.301 bytes('AA5501020101BF04')
15:43:25.303 Short press zone: 1
15:43:25.309 MQT: tele/Sonoff-Lichtschalter-Keller/RESULT = {"TXUltimate":{"Action":"Short","Channel":1}}

I think somehow there is a problem with the channel evaluation.
So if I specify in the rule at which touch position (channel 1-10) the relay should be switched.
Or am I wrong here? Or is it a simple syntax error?

Thank you in advance!

**Comment by sfromis** [comment]
Did you by any chance activate one-shot? You can see full rule set status using the command `Rule0`

Anyway, due to the contortions with rules for processing such payloads, I'd recommend to use Berry instead, providing significantly better processing options than rules. Simplistic example:
```
tasmota.add_rule('TXUltimate#Action=Short',
  def (value, trigger, payload)
    var channel=payload['TXUltimate'].find('Channel', 5)  # failsafe if no Channel
    if channel < 5 tasmota.cmd('Power1 toggle') end
    if channel > 5 tasmota.cmd('Power2 toggle') end
  end)
```

**Reply by darkdan2000** [reply]
Thanks a lot,

that works (a little bit). I've deact. the normal rules

11:50:13.923 CMD: rule0
11:50:13.929 MQT: stat/Sonoff-Lichtschalter-Keller/RESULT = {"Rule1":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
11:50:13.941 MQT: stat/Sonoff-Lichtschalter-Keller/RESULT = {"Rule2":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}
11:50:13.951 MQT: stat/Sonoff-Lichtschalter-Keller/RESULT = {"Rule3":{"State":"OFF","Once":"OFF","StopOnError":"OFF","Length":0,"Free":511,"Rules":""}}

 and added your code to the berry console.
I've now the problem that the relay switch 3 times.
any idea?

**11:41:03.755 bytes('AA5501020200012874')
11:41:03.757 Touch event: 0 pos: 1
11:41:03.763 MQT: tele/Sonoff-Lichtschalter-Keller/RESULT = {"TXUltimate":{"Action":"Touch","Channel":1}}
11:41:04.155 bytes('AA5501020101BF04')
11:41:04.157 Short press zone: 1
11:41:04.162 MQT: tele/Sonoff-Lichtschalter-Keller/RESULT = {"TXUltimate":{"Action":"Short","Channel":1}}
11:41:04.173 MQT: stat/Sonoff-Lichtschalter-Keller/RESULT = {"POWER1":"ON"}
11:41:04.177 MQT: stat/Sonoff-Lichtschalter-Keller/POWER1 = ON
11:41:04.189 MQT: stat/Sonoff-Lichtschalter-Keller/RESULT = {"POWER1":"OFF"}
11:41:04.191 MQT: stat/Sonoff-Lichtschalter-Keller/POWER1 = OFF
11:41:04.201 MQT: stat/Sonoff-Lichtschalter-Keller/RESULT = {"POWER1":"ON"}
11:41:04.205 MQT: stat/Sonoff-Lichtschalter-Keller/POWER1 = ON

11:41:06.634 bytes('AA55010202000A991F')
11:41:06.635 Touch event: 0 pos: 10
11:41:06.640 MQT: tele/Sonoff-Lichtschalter-Keller/RESULT = {"TXUltimate":{"Action":"Touch","Channel":10}}
11:41:06.763 bytes('AA550102010A0E6F')
11:41:06.765 Short press zone: 10
11:41:06.771 MQT: tele/Sonoff-Lichtschalter-Keller/RESULT = {"TXUltimate":{"Action":"Short","Channel":10}}
11:41:06.779 MQT: stat/Sonoff-Lichtschalter-Keller/RESULT = {"POWER2":"ON"}
11:41:06.783 MQT: stat/Sonoff-Lichtschalter-Keller/POWER2 = ON
11:41:06.793 MQT: stat/Sonoff-Lichtschalter-Keller/RESULT = {"POWER2":"OFF"}
11:41:06.796 MQT: stat/Sonoff-Lichtschalter-Keller/POWER2 = OFF
11:41:06.806 MQT: stat/Sonoff-Lichtschalter-Keller/RESULT = {"POWER2":"ON"}
11:41:06.808 MQT: stat/Sonoff-Lichtschalter-Keller/POWER2 = ON**

**Reply by sfromis** [reply]
Well, what I posted was an illustrative example, not "use this". As you can see, it triggers on "anything TXUltimate", meaning both Touch and Short. You may want a trigger like `TXultimate#Action=Short` instead. The trigger syntax is very similar to "classic" rules.

BTW, if you enable `Weblog 3`, you'll get an indication of the source of the action (`SRC:`), which in some situation (not so much here) is useful for debugging.
